### PR TITLE
Added 'Labels' key to testament spec

### PR DESCRIPTION
**Pull request by canelhasmateus**
Love for the test suite

Continuing with the theme of enhancing our test suite,
this commits adds another key to the testament spec.

The `labels` key is a space-separated string of metadata
describing the test, with purpose of being used
during development.

Since a full run of the test suite takes a long time,
it makes sense to run only a subset at a time. The
current `cat` folders does a very poor job , being
coarse and arbitrary. This key aims to lay a
foundation to fill the gaps.

We also continue with the efforts setted by the
previous commits:
. Remove almost all illegible names in files.
. Add a meaningful 'description' for the changed tests.
. Reduce the usage of 'echo' in the tests.

This is part of a bigger effort, see
https://github.com/nim-works/nimskull/issues/41.

### CODE REVIEW

File: tests/ccgbugs/tcodegen_anonymous_aggregate.nim [hunk 1]
```diff
@@ -0,0 +1,34 @@
+discard """
+targets: "c cpp"
```
Comment (saem) [inline tests/ccgbugs/tcodegen_anonymous_aggregate.nim:2]: This might be handled by defaultCategory in testament.Nim, have a look and it might default to these targets.

File: tests/ccgbugs/tcodegen_anonymous_aggregate.nim [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+labels: "backend_cpp , pragma , atomics , union"
```
Comment (saem) [inline tests/ccgbugs/tcodegen_anonymous_aggregate.nim]: Backend CPP is deriveable

**Review by saem — COMMENTED** [review]
Did a quick pass

File: tests/ccgbugs/tcodegen_anonymous_aggregate.nim (hunk 2 above)
Comment (canelhasmateus) [reply tests/ccgbugs/tcodegen_anonymous_aggregate.nim]: The way i set those up is that the labels contain the `origin` of the bug, not the backend in which it runs. Is this a meaningful  or useful differentiation? not sure. will remove for now

File: tests/ccgbugs/tcodegen_anonymous_aggregate.nim (hunk 1 above)
Comment (canelhasmateus) [reply tests/ccgbugs/tcodegen_anonymous_aggregate.nim:2]: the standardTargets is defined as either c or cpp , but not both. 
The defaultTargets func does not seen to cover this case, unless i'm mistaken.

**Comment by canelhasmateus** [comment]
PTAL

File: tests/actiontable/tactiontable.nim [hunk 3]
```diff
@@ -33,5 +30,5 @@ const
     "C": action3,
     "D": action4}.toTable
 
-actionTable1["C"]("arg")
-actionTable2["C"]("arg")
+doAssert actionTable1["C"]("arg") == "action 3 arg"
```
Comment (canelhasmateus) [inline tests/actiontable/tactiontable.nim]: I probably should change these names to actionTableVar and actionTableConst

File: tests/alias/talias.nim [hunk 4]
```diff
@@ -1,4 +1,9 @@
-# Test the alias analysis
+discard """
+labels: "alias"
+description: '''
+  . Test the alias analysis
```
Comment (canelhasmateus) [inline tests/alias/talias.nim]: Wasn't able to gather much information about this one.

File: tests/align/talign.nim [hunk 5]
```diff
@@ -39,15 +43,13 @@ proc foobar() =
   doAssert (cast[uint](addr(mylocal2)) and 127) == 0
   doAssert (cast[uint](addr(mylocal3)) and 127) == 0
 
-  echo "align ok"
```
Comment (canelhasmateus) [inline tests/align/talign.nim]: This echo is completely useless right? it is already preceded by doAsserts, and is not dependent on any computed value.

File: tests/align/talign.nim [hunk 6]
```diff
@@ -39,15 +43,13 @@ proc foobar() =
...
-
 foobar()
 
-# bug #13122
+block issue_13122:
```
Comment (canelhasmateus) [inline tests/align/talign.nim]: Put this second test inside a block. 
Sometimes i'm worried this changes the semantics, since top-level seems to be treated differently, but it doesn't look like the case here.

File: tests/ambsym/mresolution_identifier_1.nim [hunk 7]
```diff
@@ -0,0 +1,9 @@
+import mresolution_identifier_2
```
Comment (canelhasmateus) [inline tests/ambsym/mresolution_identifier_1.nim]: The changes to the `ambsym` category are the biggest i've made so far - i'm getting more confident in these changes. 
So please, take a more careful look at these.

File: tests/ambsym/tresolution_identifier_2_ok.nim [hunk 8]
```diff
@@ -0,0 +1,17 @@
+discard """
+labels: "enum identifier proc resolution"
+description: '''
+  . mresolution_identifier1 defines TExport, an enum with x, y, z.
+  . mresolution_identifier2 defines TExport, an enum with a, b, c.
+  . By expliciting a namespace alongside the type name, we are able to
+    resolve the ambiguity.
+  . This goes for both the TExport and the `foo` call. The `foo` call should
+    be resolved to the one defined in identifier_2, *not* in identifier_2_copy.
```
Comment (canelhasmateus) [inline tests/ambsym/tresolution_identifier_2_ok.nim]: adding a `foo` proc to mresolution_identifier_1 seems like a good idea. 
Anyway, this whole category reads like a spec, and is probably made redundant by those. Is this the case, @haxscramper ?

File: tests/ambsym/tresolution_identifier_local.nim [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+var
+  m: TMyType
+
+const size = 7
+const content = "1234"
+
+m.len = size
+m.data = content
+
+doAssert doLen(m, 5) == size
```
Comment (canelhasmateus) [inline tests/ambsym/tresolution_identifier_local.nim]: I'm not too keen on what was being tested here to begin with. 
Is it the overloading of the `len` stdlib proc to our own type? or is it the disambiguation of the x.len (field) vs x.len (UFCS)?
Why does `doLen` even has a second argument?
Does the {.inline.} pragma change the semantics?

**Review by canelhasmateus — COMMENTED** [review]
More changes. This is the deepest one i made so far, so please take a more careful look at the changes for the `ambsym` category. 
In my mind, these changes wouldy be better made in a second PR, but i didn't want to open a new one before closing this. 
Not sure if pulverizing these incoming changes over multiple PRs is better / worse for reviewing - please share your opinions.

**Comment by canelhasmateus** [comment]
pinging @zerbina into reviews as requested :P

File: tests/alias/talias.nim (hunk 4 above)
Comment (zerbina) [reply tests/alias/talias.nim]: It's an internal test for the alias analysis in `sem/aliases.nim` using the undocumented `IsPartOf` magic.

File: tests/align/talign.nim (hunk 5 above)
Comment (zerbina) [reply tests/align/talign.nim]: Besides it's usage as a progress indicator during debugging, I don't think it serves any purposes, so it's removal is fine.

File: tests/align/talign.nim (hunk 6 above)
Comment (zerbina) [reply tests/align/talign.nim]: A `block:` (`static:` too) only opens a new scope, the content is still treated as top-level.

Top-level variables (globals) have various differences in semantics compared to local variables. For example, they're not captured in closures and are copied into `sink` arguments (instead of moved) by default.

Comment (canelhasmateus) [reply tests/align/talign.nim]: TIL

**Comment by zerbina** [comment]
> pinging @zerbina into reviews as requested :P

Review request accepted :D

> In my mind, these changes wouldy be better made in a second PR, but i didn't want to open a new one before closing this.
> Not sure if pulverizing these incoming changes over multiple PRs is better / worse for reviewing - please share your opinions.

I think splitting the `ambsym` category changes into a separate PR would be a good idea too. Smaller and separate PRs often allow for changes to be merged earlier

**Comment by canelhasmateus** [comment]
> I think splitting the `ambsym` category changes into a separate PR would be a good idea too. Smaller and separate PRs often allow for changes to be merged earlier

Done. Removed the second commit from this one, and opened a second PR.
Do you mind reviewing / borsing this one @zerbina ?

File: tests/ccgbugs/tcodegen_bare_return.nim [hunk 10]
```diff
@@ -0,0 +1,30 @@
+discard """
+labels: "codegen error_runtime exception"
+description: '''
+  . From https://github.com/nim-lang/Nim/issues/9286
+    Exception doesn't clear result value
+  . The second time should return none, but isSome(opt) is always equals to true
+  . Wrong heuristic in codegen, A bare return may trigger the insertion
```
Comment (zerbina) [inline tests/ccgbugs/tcodegen_bare_return.nim]:
```suggestion
  . Wrong heuristic in codegen. A bare return may trigger the insertion
```

File: tests/ccgbugs/tcodegen_char_cast_overflow.nim [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+
+let c = '\128'
+
+block case1:
+  doAssert data[c.int] == 'y'
+  doAssert seq2[c.int] == 'y'
+
+
+block case2:
+   proc play(x: openArray[char]) =
```
Comment (zerbina) [inline tests/ccgbugs/tcodegen_char_cast_overflow.nim]:
```suggestion
  proc play(x: openArray[char]) =
```

File: tests/ccgbugs/tcodegen_distinct_uint64_32bit.nim [hunk 12]
```diff
@@ -0,0 +1,20 @@
+discard """
+labels: "32_bit distinct var"
+description: '''
+  . From https://github.com/nim-lang/Nim/issues/13902
+    distinct uint64" type corruption on 32-bit, when using {.borrow.} operators
```
Comment (zerbina) [inline tests/ccgbugs/tcodegen_distinct_uint64_32bit.nim]:
```suggestion
    "distinct uint64" type corruption on 32-bit, when using {.borrow.} operators
```

File: tests/ccgbugs/tcodegen_generic_openarray_import.nim [hunk 13]
```diff
@@ -10,6 +10,13 @@ output: '''
 @[(v: -1), (v: 2), (v: 3)]
 @[(v: -1), (v: 2), (v: 3)]
 '''
+labels : "codegen generic import openArray"
+description: '''
+  . From https://github.com/nim-lang/Nim/issues/9578
+    openArray generates incorrect C code with "incomplete type"
+  . OpenArray[T] generates incorrect C code, if the containing module
```
Comment (zerbina) [inline tests/ccgbugs/tcodegen_generic_openarray_import.nim]:
```suggestion
  . openArray[T] generates incorrect C code, if the containing module
```

File: tests/ccgbugs/tcodegen_release_gcc_volatile.nim [hunk 14]
```diff
@@ -0,0 +1,23 @@
+discard """
+matrix: "-d:release"
+labels: "code_gen gcc mode_release"
```
Comment (zerbina) [inline tests/ccgbugs/tcodegen_release_gcc_volatile.nim]:
```suggestion
labels: "code_gen mode_release"
```

Not related to `gcc`.

File: tests/ccgbugs/tcodegen_signed_zero.nim [hunk 15]
```diff
@@ -0,0 +1,16 @@
+discard """
+targets: "c js"
+labels: "backend_VM codegen floating_point"
```
Comment (zerbina) [inline tests/ccgbugs/tcodegen_signed_zero.nim]:
```suggestion
targets: "c js"
labels: "codegen floating_point"
```

This issue is not related to the VM.

File: tests/ccgbugs/tcodegen_template_with_generic.nim [hunk 16]
```diff
@@ -0,0 +1,30 @@
+discard """
+labels: "codegen error_compilation generics template"
+description: '''
+  . From https://github.com/nim-lang/Nim/issues/6756
+    Error in template when using the type of the parameter inside it
+  . The problem is in the line var res: A[T]. The assert fails because symbol is
+    A_105025: ["sfUsed"], [], ["issue.nim", 5, 15], skType
```
Comment (zerbina) [inline tests/ccgbugs/tcodegen_template_with_generic.nim]:
```suggestion
  . The problem is in the line `var res: A[T]`. The assert fails because symbol is
    `A_105025: ["sfUsed"], [], ["issue.nim", 5, 15], skType`
```

File: tests/ccgbugs/tgc_bug.nim [hunk 17]
```diff
@@ -0,0 +1,16 @@
+discard """
+matrix: "--gc:refc -d:fulldebug -d:smokeCycles"
+output: '''@["a"]'''
+labels: "32_bit_only codegen gc gcc seq"
```
Comment (zerbina) [inline tests/ccgbugs/tgc_bug.nim]:
```suggestion
labels: "codegen gc seq"
```

The issue is not related to the C compiler or processor architecture.

File: tests/ccgbugs/xarray9578.nim [hunk 18]
```diff
@@ -1,7 +1,8 @@
-import t9578
+import tcodegen_generic_openarray_import
 
 proc testArray*(x: var array[3,mytype]) =
   f(x[0].addr)
 
 proc testArray2*(x: var ptr array[3,mytype]) =
   f(x[0].addr)
+
```
Comment (zerbina) [inline tests/ccgbugs/xarray9578.nim:8]: Some files end with a new-line and some don't. I think this should be done consistently, but I'm not sure which style is preferred.

**Review by zerbina — CHANGES_REQUESTED** [review]
Some labels don't apply, but after that's fixed I'm good to merge.

I also left some suggestions for improved readability, but these are non-blocking.

**Comment by zerbina** [comment]
Also, could you remove the template `Summary`/`Details` part from the PR message body?

File: tests/ccgbugs/tgc_bug.nim (hunk 17 above)
Comment (canelhasmateus) [reply tests/ccgbugs/tgc_bug.nim]: In my [defense](https://github.com/nim-lang/Nim/issues/6279#issuecomment-324914791) :P

**Comment by canelhasmateus** [comment]
There you go

**Review by zerbina — APPROVED** [review]
Nice, thank you. Changing the final new-line style (if wanted, that is) can be done in a separate PR.

bors r+

**Comment by canelhasmateus** [comment]
This timeout is probably my fault. By accepting each of your suggestions separately i inadvertently kicked off multiple CI runs

**Comment by canelhasmateus** [comment]
Very nice! Thank you for your attention.
